runners/direct-java/src/main/java/org/apache/beam/runners/direct/AggregatorContainer.java

```diff
@@ -0,0 +1,159 @@
+    private AccumT accumulator = null;
+    private boolean committed = false;
+
+    private AggregatorInfo(String name, CombineFn<InputT, AccumT, OutputT> combiner) {
+      this.name = name;
+      this.combiner = combiner;
+    }
+
+    @Override
+    public synchronized void addValue(InputT input) {
+      Preconditions.checkState(!committed, "Cannot addValue after committing");
+      if (accumulator == null) {
+        accumulator = combiner.createAccumulator();
+      }
+      accumulator = combiner.addInput(accumulator, input);
+    }
+
+    public synchronized OutputT getOutput() {
+      return accumulator == null ? null : combiner.extractOutput(accumulator);
+    }
+
+    private void merge(AggregatorInfo<?, ?, ?> other) {
+      // Aggregators are only merged if they are the same (same step, same name).
+      // As a result, they should also have the same CombineFn, so this is safe.
+      AggregatorInfo<InputT, AccumT, OutputT> otherSafe =
+          (AggregatorInfo<InputT, AccumT, OutputT>) other;
+      mergeSafe(otherSafe);
+    }
+
+    private synchronized void mergeSafe(AggregatorInfo<InputT, AccumT, OutputT> other) {
+      if (accumulator == null) {
+        accumulator = other.accumulator;
+      } else if (other.accumulator != null) {
+        accumulator = combiner.mergeAccumulators(Arrays.asList(accumulator, other.accumulator));
+      }
+    }
+
+    @Override
+    public String getName() {
+      return name;
+    }
+
+    @Override
+    public CombineFn<InputT, ?, OutputT> getCombineFn() {
+      return combiner;
+    }
+  }
+
+  private final ConcurrentHashMap<String, AggregatorInfo<?, ?, ?>> accumulators =
+      new ConcurrentHashMap<>();
+
+  private AggregatorContainer() {
+  }
+
+  public static AggregatorContainer create() {
+    return new AggregatorContainer();
+  }
+
+  @Nullable
+  <OutputT> OutputT getAggregate(String name) {
+    AggregatorInfo<?, ?, OutputT> aggregatorInfo =
+        (AggregatorInfo<?, ?, OutputT>) accumulators.get(name);
+    return aggregatorInfo == null ? null : aggregatorInfo.getOutput();
+  }
+
+  public Mutator createMutator() {
+    return new Mutator(this);
+  }
+
+  /**
+   * AccumT class for mutations to the aggregator values.
+   */
+  public static class Mutator implements AggregatorFactory {
+
+    private final Map<String, AggregatorInfo<?, ?, ?>> accumulatorDeltas = new HashMap<>();
+    private final AggregatorContainer container;
+    private boolean committed = false;
+
+    private Mutator(AggregatorContainer container) {
+      this.container = container;
+    }
+
+    public void commit() {
+      Preconditions.checkState(!committed, "Should not be already committed");
+      committed = true;
+
+      for (Map.Entry<String, AggregatorInfo<?, ?, ?>> entry : accumulatorDeltas.entrySet()) {
+        AggregatorInfo<?, ?, ?> previous = container.accumulators.get(entry.getKey());
+        entry.getValue().committed = true;
+        if (previous == null) {
```

Comment:
Done I think? Not convinced it adds value in this case, since even if it uncovers problems it will do so in a flaky manner...